core/src/entry_verification_service/worker.rs

```diff
@@ -0,0 +1,524 @@
+
+                    let data_set_start = DataSetFirstShred::from(*shard_indices.start());
+
+                    let parent_slot = commands::add_data_set(
+                        &mut todo,
+                        blockstore,
+                        processing,
+                        waiting_for_parent_hash,
+                        results,
+                        slot,
+                        shard_indices,
+                        last_in_slot,
+                        entries,
+                    );
+
+                    // If the client is waiting on the slot and data set that we have work for,
+                    // stop the work accumulation.
+                    //
+                    // Same for any work in the last data set of the parent slot, as the first child
+                    // slot is blocked by the parent slot work.
+                    if let Some((waiting_slot, waiting_data_set_start)) = client_waiting_for {
+                        let last_data_set_in_parent = parent_slot
+                            .map(|parent_slot| last_in_slot && parent_slot == *waiting_slot)
+                            .unwrap_or(false);
+
+                        if last_data_set_in_parent
+                            || (slot == *waiting_slot && data_set_start <= *waiting_data_set_start)
+                        {
+                            return Some(todo);
+                        }
+                    }
+                }
+                Command::WaitingFor { slot, data_set } => {
+                    match commands::waiting_for(&todo, client_waiting_for, slot, data_set) {
+                        commands::AccumulationDecision::RunVerification => return Some(todo),
+                        commands::AccumulationDecision::CanWaitMore => (),
+                    }
+                }
+                Command::Cleanup { earlier_than } => commands::cleanup(
+                    processing,
+                    waiting_for_parent_hash,
+                    client_waiting_for,
+                    results,
+                    earlier_than,
+                ),
+            }
+        }
+    }
+
+    fn verify(&mut self, todo: Todo) {
+        let Self {
+            processing,
+            results,
+            ..
+        } = self;
+
+        // As we do not expect to process more than 1 or 2 slots at a time, we use a vector, rather
+        // than a set to store slot indices.
+        let mut verified = vec![];
+        for TodoItem { slot, .. } in todo.iter() {
+            if !verified.contains(slot) {
+                verified.push(*slot);
+            }
+        }
+
+        // TODO Send request to the GPU thread.
+
+        // TODO Run signature verification.
+
+        let failed = vec![];
+        // TODO Check results and mark any failed slots.
+
+        for slot in failed.iter() {
+            if let Some(i) = verified.iter().position(|v| *v == *slot) {
+                verified.swap_remove(i);
+            }
```

Comment:
Would None here be considered an error?

Reply:
It would.  But I was trying to avoid `unwrap()` and `panic!()` calls.
`unexpected_error!` will just report an error in a release build and continue.
Here, `None` should never happen, yes.  It would be a logic error in the `verify()` call logic.
I can change `if` to `expect()`.  But it does not seem to be that bad that we want to stop the validator.  Not sure.
It should also be checked by unit tests anyway.  I will write some for this method when it will be connected to an actual verification call.